src/providers/ows/qgsgeocmsdataitems.cpp

```diff
@@ -0,0 +1,270 @@
+      skipProvider = true;
+      continue;
+    }
+
+    QString path =  pathPrefix + mName;
+    QgsDebugMsg( "path = " + path );
+
+    QVector<QgsDataItem *> items;
+    Q_FOREACH ( QgsDataItemProvider *pr, dataItemProvidersFn() )
+    {
+      items = pr->name().startsWith( mServiceName ) ? pr->createDataItems( path, this ) : items;
+      if ( !items.isEmpty() )
+      {
+        break;
+      }
+    }
+
+    if ( items.isEmpty() )
+    {
+      QgsDebugMsg( "Connection not found by provider" );
+      skipProvider = true;
+      continue;
+    }
+
+    if ( mServiceName == QStringLiteral( "XYZ" ) )
+    {
+      QgsDebugMsg( "Add new item : " + mServiceName );
+      return items;
+    }
+
+    Q_FOREACH ( QgsDataItem *item, items )
+    {
+      item->populate( true ); // populate in foreground - this is already run in a thread
+
+      layerCount += item->rowCount();
+      if ( item->rowCount() > 0 )
+      {
+        QgsDebugMsg( "Add new item : " + item->name() );
+        serviceItems.insert( item, key );
+      }
+      else
+      {
+        //delete item;
+      }
+    }
+
+    skipProvider = true;
+  }
+
+  Q_FOREACH ( QgsDataItem *item, serviceItems.keys() )
+  {
+    QgsDebugMsg( QString( "serviceItems.size = %1 layerCount = %2 rowCount = %3" ).arg( serviceItems.size() ).arg( layerCount ).arg( item->rowCount() ) );
+    QString providerKey = serviceItems.value( item );
+
+    // Add layers directly to service item
+    Q_FOREACH ( QgsDataItem *subItem, item->children() )
+    {
+      if ( subItem->path().endsWith( QString( "error" ) ) )
+      {
+        continue;
+      }
+      item->removeChildItem( subItem );
+      subItem->setParent( this );
+      replacePath( subItem, providerKey.toLower() + ":/", pathPrefix );
+      children.append( subItem );
+    }
+
+    delete item;
+  }
+
+  return children;
+}
+
+// reset path recursively
+void QgsGeoCMSServiceItem::replacePath( QgsDataItem *item, QString before, QString after )
+{
+  item->setPath( item->path().replace( before, after ) );
+  Q_FOREACH ( QgsDataItem *subItem, item->children() )
+  {
+    replacePath( subItem, before, after );
+  }
+}
+
+QgsGeoCMSRootItem::QgsGeoCMSRootItem( QgsDataItem *parent, QString name, QString path ) : QgsDataCollectionItem( parent, name, path )
+{
+  mCapabilities |= Fast;
+  if ( name == QString( "GeoNode" ) )
+  {
+    mIconName = QStringLiteral( "mIconGeonode.svg" );
+  }
+  else
+  {
+    mIconName = QStringLiteral( "mIconConnect.svg" );
+  }
+  populate();
+}
+
+QVector<QgsDataItem *> QgsGeoCMSRootItem::createChildren()
+{
+  QVector<QgsDataItem *> connections;
+
+  Q_FOREACH ( const QString &connName, QgsGeoCmsConnection::connectionList( mName ) )
+  {
+    QgsGeoNodeConnection *connection = nullptr;
+    if ( mName == QString( "GeoNode" ) )
+    {
+      connection = new QgsGeoNodeConnection( connName );
+    }
+    else
+    {
+      connection = new QgsGeoNodeConnection( connName );
```

Comment:
the created "connection" will leak memory. Here it is best to create the "connection" objects on stack rather than on heap (i.e. ```QgsGeoNodeConnection connection( connName );``` ) and pass that as a reference to QgsCMSConnectionItem's constructor (which can make a copy).

Also it seems if/else here is maybe not necessary ;-)